src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Stream.cs

```diff
@@ -39,14 +39,24 @@ public static Task SerializeAsync<TValue>(Stream utf8Json, TValue value, JsonSer
         public static Task SerializeAsync(Stream utf8Json, object? value, Type inputType, JsonSerializerOptions? options = null, CancellationToken cancellationToken = default)
         {
             if (utf8Json == null)
+            {
                 throw new ArgumentNullException(nameof(utf8Json));
+            }
+
+            if (inputType == null)
+            {
+                throw new ArgumentNullException(nameof(inputType));
+            }
 
-            VerifyValueAndType(value, inputType);
+            if (value != null && !inputType.IsAssignableFrom(value.GetType()))
+            {
+                ThrowHelper.ThrowArgumentException_DeserializeWrongType(inputType, value);
+            }
 
-            return WriteAsyncCore(utf8Json, value, inputType, options, cancellationToken);
+            return WriteAsyncCore<object>(utf8Json, value!, inputType, options, cancellationToken);
```

Comment:
This is incorrect usage of nullability annotation. If value can be null (and looking at the null checks it can), the `TValue value` parameter on `WriteAsyncCore` should be correctly annotated to be nullable.

Avoid using `!` to circumvent null warnings unless absolutely necessary, and instead annotate the implementation to reflect nullability accurately.

Reply:
I agree. This should be called with `WriteAsyncCore<object?>`.

Reply:
Maybe a helpful note for reviewers: Make sure to pay close attention to uses of `!` in the code :)